### ISSUE TITLE
Add cached task names to `fractal workflow add-task`

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal#309]
This is currently only available in the `fractal task` subcommands.

**Comment by jluethi** [comment]
When using the client, it would be very convenient to select a task to add to a workflow by name & version when adding them with `fractal workflow add-task`. And would be even nicer to be have the version optional with a way to default to the newest version, unless specified otherwise.

Currently, we select them by ID. But when we have multiple versions of tasks around, I need to list them to get their IDs. And those aren't fully reproducible (depends on what order the different versions of the tasks were installed and how many versions are installed for a given server installation) => tricky for examples

**Comment by tcompa** [comment]
> When using the client, it would be very convenient to select a task to add to a workflow by name & version when adding them with `fractal workflow add-task`. And would be even nicer to be have the version optional with a way to default to the newest version, unless specified otherwise.

This feature can be included somewhat easily, if we restrict its scope to a subset of all possible tasks - namely the ones coming from a pypi install of `fractal-tasks-core` package. This would mean that:
1. One of these tasks can be addressed by:
    * ID
    * name + version (version can be optional and will default to latest)
2. Any other task can only be addressed by ID.

Note that restricting the scope to this subset corresponds to a filter on the relevant task `source` attributes, which will only include the tasks with sources like `pip:fractal-tasks-core==1.2.3` (or similar, if we change the syntax).

@jluethi, thoughts?

**Comment by tcompa** [comment]
Addendum to the last comment: this would imply that only core tasks (installed from pypi) are stored in the cache, which is then a dictionary mapping `(name,version)` (internally encoded as `source`) to `id`.

**Comment by jluethi** [comment]
> One of these tasks can be addressed by:
> - ID
> - name + version (version can be optional and will default to latest)

I think this form of addressing it makes sense for core tasks. 

As mentioned elsewhere, ID is not a portable or reproducible info, because it depends on how things were installed on the current server. Thus, if we want to enable running examples on different servers or downloading of workflows & uploading them to a different server / a restarted server / a server with multiple versions of the core tasks, then the matching needs to happen via name + source + version.

We can default to core tasks being addressable as name + version, as this should be the most common type of task being added.

But there would need to be some way to specify tasks from other sources that isn't their ID, at least when uploading workflows. And if we have that way, would be (eventually) useful to expose it to the client as well.

**Comment by tcompa** [comment]
> But there would need to be some way to specify tasks from other sources that isn't their ID, at least when uploading workflows. And if we have that way, would be (eventually) useful to expose it to the client as well.

This is precisely the goal of the `source` attribute. At least for pypi-installed tasks, it provides a global (across Fractal-sever instances) unique identifier.

TL;DR more discussion is needed (here or somewhere else) about the identity of tasks in more general cases, and then we can adapt the client accordingly

---

There is no way (at least at the moment) to uniquely identify a task in the other cases (i.e. apart from the pypi case). If a user adds a custom task, the source will be something on the line of `custom-task-by-user-X-some-more-info`, possibly with its own version or other info in there. This task can be part of a workflow, and exporting a workflow will also export the task properties. Up to here everything is smooth.

Now, importing this task in a different Fractal instance is extremely unreliable. The task with source `custom-task-by-user-X-some-more-info` could be present or missing. If it is missing, just importing it would not be sufficient because the `command` attribute is specific to the Fractal instance (e.g. it includes executable paths), and cannot simply be copy-pasted. And even if it is present there is nothing that guarantees that it is the same task as in the original Fractal instance (the database is obviously not shared across instances).

The only trivial way to export a task (as part of a workflow) and re-import it somewhere else is to copy-paste all of its files, which may work for very simple tasks (a python file with very slim dependencies) but is obviously not general (what would we do if a task is just a wrapper of a heavy external non-python package? no way we can clone that..). And it would also mean that we define a list of task-related files, which doesn't seem a good option at all..

**Comment by jluethi** [comment]
> There is no way (at least at the moment) to uniquely identify a task in the other cases (i.e. apart from the pypi case)

I think that limitation is fair enough! If one wants to share tasks between repos, they should be installable as a python package, thus have a reproducible source.
We may get to the point where we need to support more than pypi (to allow someone to have closed-source tasks to be run within Fractal), but I don't think we need to concern ourselves with handling dev builds of tasks.

The point above was more that we may need to support tasks installed from other pypi packages with more than the ID. If custom tasks need to be addressed by ID, I'm fine with that limit to the scope.

**Comment by tcompa** [comment]
The limited-scope of this feature is available via #341. If there are no name clashes (e.g. if a single version of tasks is installed), addressing tasks by name should now work also in `fractal workflow add-task`. If there are non-unique names, the cache is never written and only IDs can be used.

This is available for "real-life" testing with client version `1.0.0a1` (on pypi).

**Comment by tcompa** [comment]
Re-opening due to this error
```
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:root:Namespace(args_file='Parameters/args_create_ome_zarr.json', batch=False, cmd='workflow', id='1', json=False, order=None, password=None, slurm_user=None, subcmd='add-task', task_id='Create OME-Zarr structure', user=None, v=0)
Traceback (most recent call last):
  File "/data/homes/test01/.conda/envs/fractal-client-v1/bin/fractal", line 8, in <module>
    sys.exit(run())
  File "/data/homes/test01/.conda/envs/fractal-client-v1/lib/python3.8/site-packages/fractal/__main__.py", line 7, in run
    asyncio.run(main())
  File "/data/homes/test01/.conda/envs/fractal-client-v1/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/data/homes/test01/.conda/envs/fractal-client-v1/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/data/homes/test01/.conda/envs/fractal-client-v1/lib/python3.8/site-packages/fractal/client/client.py", line 65, in main
    interface = await handle()
  File "/data/homes/test01/.conda/envs/fractal-client-v1/lib/python3.8/site-packages/fractal/client/client.py", line 55, in handle
    interface = await handler(client, **vars(args))
  File "/data/homes/test01/.conda/envs/fractal-client-v1/lib/python3.8/site-packages/fractal/client/cmd/__init__.py", line 161, in workflow
    iface = await workflow_add_task(client, **kwargs)
  File "/data/homes/test01/.conda/envs/fractal-client-v1/lib/python3.8/site-packages/fractal/client/cmd/_workflow.py", line 91, in workflow_add_task
    workflow_task = WorkflowTaskCreate(task_id=task_id, order=order)
  File "/data/homes/test01/.conda/envs/fractal-client-v1/lib/python3.8/site-packages/sqlmodel/main.py", line 506, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for WorkflowTaskCreate
task_id
  value is not a valid integer (type=type_error.integer)
```

**Comment by tcompa** [comment]
The last error was related to using the wrong client version. With `1.0.0a1` we now have a working example 01 in the demos (as of https://github.com/fractal-analytics-platform/fractal-demos/commit/530c3736a66cc7bad7a098f069144c08197bc2cb).
